src/aws-cpp-sdk-core/source/auth/AWSCredentialsProviderChain.cpp

```diff
@@ -21,11 +21,20 @@ static const char DefaultCredentialsProviderChainTag[] = "DefaultAWSCredentialsP
 
 AWSCredentials AWSCredentialsProviderChain::GetAWSCredentials()
 {
+    std::unique_lock<std::mutex> lock(m_providerChainLock);
```

Comment:
Use semaphore class available in the SDK
(multi-read or single write)

Reply:
Took a look at `Semaphore` and saw that `ReaderLockGuard` actually uses it and provides a more convenient API for what im trying to do. we actually already use it in [credentials provider](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/auth/AWSCredentialsProvider.cpp#L204-L219) in basically the same use case.

```
void ProfileConfigFileAWSCredentialsProvider::RefreshIfExpired()
{
    ReaderLockGuard guard(m_reloadLock);
    if (!IsTimeToRefresh(m_loadFrequencyMs))
    {
       return;
    }

    guard.UpgradeToWriterLock();
    if (!IsTimeToRefresh(m_loadFrequencyMs)) // double-checked lock to avoid refreshing twice
    {
        return;
    }

    Reload();
}
```

Let me know if you're ok with that